<a href="https://colab.research.google.com/github/baugher90/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/Development/module2-make-features/LS_DS_112_Make_Features.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 2*

---

# Make Features 

- Student should be able to understand the purpose of feature engineering
- Student should be able to work with strings in pandas
- Student should be able to work with dates and times in pandas
- Student should be able to modify or create columns of a dataframe using the `.apply()` function


Helpful Links:
- [Minimally Sufficient Pandas](https://medium.com/dunder-data/minimally-sufficient-pandas-a8e67f2a2428)
- [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)
- Python Data Science Handbook
  - [Chapter 3.10](https://jakevdp.github.io/PythonDataScienceHandbook/03.10-working-with-strings.html), Vectorized String Operations
  - [Chapter 3.11](https://jakevdp.github.io/PythonDataScienceHandbook/03.11-working-with-time-series.html), Working with Time Series
- [Lambda Learning Method for DS - By Ryan Herr](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit?usp=sharing)

# [Objective](#feature-engineering) - The Purpose of Feature Engineering



## Overview

Feature Engineering is the process of using a combination of domain knowledge, creativity and the pre-existing columns of a dataset to create completely new columns.

 Machine Learning models try to detect patterns in the data and then associate those patterns with certain predictions. The hope is that by creating new columns on our dataset that we can expose our model to new patterns in the data so that it can make better and better predictions.

This is largely a matter of understanding how to work with individual columns of a dataframe with Pandas --which is what we'll be practicing today!

## Follow Along

Columns of a dataframe hold each hold a specific type of data. Lets inspect some of the common datatypes found in datasets and then we'll make a new feature on a dataset using pre-existing columns.

In [0]:
import pandas as pd

In [0]:
# Lets take a look at the Ames Iowa Housing Dataset:
df = pd.read_csv('https://raw.githubusercontent.com/ryanleeallred/datasets/master/Ames%20Housing%20Data/train.csv')
df.head()

### Specific Columns hold specific kinds of data

In [0]:
#control how many rows or colmns we see in output
pd.set_option('display.max_rows', 100)
#object types are typicaly strings but have mixed types and pandas can't tell what the main type is
df.dtypes

Some columns hold integer values like the `BedroomAbvGr` which stands for "Bedrooms Above Grade." This is the number of non-basement bedrooms in the home.

For more information on specific column meanings view the [data dictionary](https://github.com/ryanleeallred/datasets/blob/master/Ames%20Housing%20Data/data_description.txt).

In [0]:
# Look at the first ten rows of the `BedroomAbvGr` column.
# Looks like integers to me!
df['BedroomAbvGr']

Some columns hold float values like the `LotFrontage` column.

In [16]:
# Look at the first ten rows of the `LotFrontage` column. Floats
df['LotFrontage']

0       65.0
1       80.0
2       68.0
3       60.0
4       84.0
        ... 
1455    62.0
1456    85.0
1457    66.0
1458    68.0
1459    75.0
Name: LotFrontage, Length: 1460, dtype: float64

In [0]:
#object type
df['MSZoning']

Hmmm, do the values above look like floats to you?

They all have .0 on them so technically they're being stored as floats, but *should* they be stored as floats?

Lets see what all of the possible values for this column are.

In [0]:
pd.set_option('display.max_rows', 150)

#(dropna=False)includes NaN value to get acurate count of what values are in the columns
df['LotFrontage'].value_counts(dropna=False)

Looks to me like the `LotFrontage` column originally held integer values but was cast to a `float` meaning that each original integer values was converted to its corresponding float representation. 

Any guesses as to why that would have happened?


HINT: What's the most common `LotFrontage` value for this column?

In [26]:
# NaN is the most common value in this column. What is a NaN
import numpy as np
#NaN is a float and there are a lot of NaNs then it will make all other integers in the column into floats
type(np.NaN)

float

`NaN` stands stands for "Not a Number" and is the default missing value indicator with Pandas. This means there were cells in this column that didn't have a LotFrontage value recorded for those homes. 

This is where domain knowledge starts to come in. Think about the context we're working with here: houses. What might a null or blank cell representing "Linear feet of street connected to property" mean in the context of a housing dataset?

Ok, so maybe it makes seanse to have some NaNs in this column. What is the datatype of a NaN value?

Perhaps some of this data is truly missing or unrecorded data, but sometimes `NaNs` are more likely to indicate something that was "NA" or "Not Applicable" to a particular observation. There could be multiple reasons why there was no value recorded for a particular feature.

Remember, that Pandas tries to maintain a single datatype for all values in a column, and therefore...

In [0]:
import numpy as np

# What is the datatype of NaN?-Float

The datatype of a NaN is float!  This means that if we have a column of integer values, but the column has even a single `NaN` that column will not be treated with the integer datatype but all of the integers will be converted to floats in order to try and preserve the same datatype throughout the entire column.

You can see already how understanding column datatypes is crucial to understanding how Pandas help us manage our data.

### Making new Features

Lets slim down the dataset and consider just a few specific columns:

- `TotalBsmtSF`
- `1stFlrSF`
- `2ndFlrSF`
- `SalePrice1`


In [53]:
# I can make a smaller dataframe with a few specific column headers
# by passing a list of column headers inside of the square brackets
#copy to creat a new array
df_small = df[['TotalBsmtSF','1stFlrSF','2ndFlrSF','SalePrice']].copy()
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice
0          856       856       854     208500
1         1262      1262         0     181500
2          920       920       866     223500
3          756       961       756     140000
4         1145      1145      1053     250000

### Syntax for creating new columns

When making a new column on a dataframe, we have to use the square bracket syntax of accessing a column. We can't use "dot syntax" here.

In [54]:
# Lets add up all of the square footage to get a single square footage 
# column for the entire dataset

# Using bracket syntax to make a new 'TotalSquareFootage' column

df_small['TotalSF'] = df_small['TotalBsmtSF'] + df_small['1stFlrSF'] + df_small['2ndFlrSF']
                                                         
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSF
0          856       856       854     208500     2566
1         1262      1262         0     181500     2524
2          920       920       866     223500     2706
3          756       961       756     140000     2473
4         1145      1145      1053     250000     3343

In [55]:
# Lets make a nother new column that is 'PricePerSqFt' by
# dividing the price by the square footage
df_small['PricePerSqFt'] = df_small['SalePrice'] / df_small['TotalSF']
df_small.head()

TotalBsmtSF  1stFlrSF  2ndFlrSF  SalePrice  TotalSF  PricePerSqFt
0          856       856       854     208500     2566     81.254871
1         1262      1262         0     181500     2524     71.909667
2          920       920       866     223500     2706     82.594235
3          756       961       756     140000     2473     56.611403
4         1145      1145      1053     250000     3343     74.783129

Ok, we have made two new columns on our small dataset.

- What does a **high** `PricePerSqFt` say about a home that the square footage and price alone don't capture as directly?

- What does a **low** `PricePerSqFt` say about a home that the square footage and price alone don't directly capture?



## Challenge

I hope you can see how we have used existing columns to create a new column on a dataset that say something new about our unit of observation. This is what making new features (columns) on a dataset is all about and why it's so essential to data science --particularly predictive modeling "Machine Learning." 

We'll spend the rest of the lecture and assignment today trying to get as good as we can at manipulating (cleaning) and creating new columns on datasets.

# [Objective](#work-with-strings) Work with Strings with Pandas

## Overview

So far we have worked with numeric datatypes (ints and floats) but we haven't worked with any columns containing string values. We can't simply use arithmetic to manipulate string values, so we'll need to learn some more techniques in order to work with this datatype.

## Follow Along

We're going to import a new dataset here to work with. This dataset is from LendingClub and holds information about loans issued in Q4 of 2018. This dataset is a bit messy so it will give us plenty of opportunities to clean up existing columns as well as create new ones.

The `!wget` shell command being used here does exactly the same thing that your browser does when you type a URL in the address. It makes a request or "gets" the file at that address. However, in our case the file isn't a webpage, it's a compressed CSV file. 

Try copying and pasting the URL from below into your browser, did it start an automatic download? Any URLs like this that start automatic downloads when navigated to can be used along with the `!wget` command to bring files directly into your notebook's memory.

### Load a new dataset

In [62]:
#for zipped files use shell command to import them from url

!wget https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip


--2020-02-19 14:41:15--  https://resources.lendingclub.com/LoanStats_2018Q4.csv.zip
Resolving resources.lendingclub.com (resources.lendingclub.com)... 64.48.1.20
Connecting to resources.lendingclub.com (resources.lendingclub.com)|64.48.1.20|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘LoanStats_2018Q4.csv.zip’

LoanStats_2018Q4.cs     [           <=>      ]  21.78M   393KB/s    in 58s     

2020-02-19 14:42:14 (386 KB/s) - ‘LoanStats_2018Q4.csv.zip’ saved [22838879]



We need to use the `!unzip` command to extract the csv from the zipped folder.

In [63]:
!unzip LoanStats_2018Q4.csv.zip

Archive:  LoanStats_2018Q4.csv.zip
  inflating: LoanStats_2018Q4.csv    


We can also use bash/shell commands to look at the raw file using the `!head` and `!tail` commands

In [0]:
!head LoanStats_2018Q4.csv

In [0]:
!tail LoanStats_2018Q4.csv

As we look at the raw file itself, do you see anything that might cause us trouble as we read in the CSV file to a dataframe?

In [0]:
import pandas as pd
pd.set_option('display.max_rows',150)

In [0]:
# Read in the CSV
df = pd.read_csv('LoanStats_2018Q4.csv')
df.head()

The extra rows at the top and bottom of the file have done two things:

1) The top row has made it so that the entire dataset is being interpreted as column headers

2) The bottom two rows have been read into the 'id' column and are causing every column to have at least two `NaN` values in it.

In [0]:
# We can fix the header problem by using the 'skiprows' parameter
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1)
print(df.shape)
df.head()

Lets look at the NaN values of each column so that you can see the problem that the extra rows at the bottom of the file are creating for us

In [0]:
# Sum null values by column and sort from least to greatest

In [0]:
# Address the extra NaNs in each column by skipping the footer as well.
df = pd.read_csv('LoanStats_2018Q4.csv', skiprows=1,skipfooter=2)
print(df.shape)
df.head()

In [0]:
df.isnull().sum().sort_values()

For good measure, we'll also drop some columns that are made up completely of NaN values.

Why might LendingClub have included columns in their dataset that are 100% blank?

In [0]:
df = df.drop(['url','member_id','desc','id'], axis=1)
df.head()

### Clean up the `int_rate` column

When we're preparing a dataset for a machine learning model we typically want to represent don't want to leave any string values in our dataset --because it's hard to do math on words. 

Specifically, we have a column that is representing a numeric value, but currently doesn't have a numeric datatype. Lets look at the first 10 values of the `int_rate` column

In [0]:
# Look at the first 10 values of the int_rate column
df['int_rate'].head(10)

In [0]:
# Look at a specific value from the int_rate column
df['int_rate'][:9]

In [0]:
df['int_rate'].describe()

Problems that we need to address with this column:

- String column that should be numeric
- Percent Sign `%` included with the number
- Leading space at the beginning of the string

However, we're not going to try and write exactly the right code to fix this column in one go. We're going to methodically build up to the code that will help us address these problems.


In [89]:
# Lets start with just fixing a single string.
# If we can fix one, we can usually fix all of them
int_rate = " 13.56%"

'13.56%'

In [0]:
#default .strip removes white space on either side
int_rate.strip()

In [0]:
#default .strip removes white space on either side
int_rate.strip("%")

In [0]:
#default .strip removes white space on either side
int_rate.strip().strip("%")

In [91]:
# "Cast" the string value to a float
float('8.5')

8.5

In [90]:
float(int_rate.strip().strip("%"))

13.56

In [0]:
float(df['int_rate'][0].strip().strip("%"))

In [0]:
#could use for loop
new_int_rate = []

for cell in df['int_rate']:
  new_int_rate.append(float(cell.strip().strip("%")))
  
new_int_rate

### Write a function to make our solution reusable!

In [0]:
# Write a function that can do what we have written above to any 
# string that is passsed to it.
def int_rate_to_float(cell_contents):
  return float(cell_contents.strip().strip("%"))

In [102]:
# Test out our function by calling it on our example
int_rate_to_float(int_rate)

13.56

In [103]:
# is the data type correct?
type(int_rate_to_float(int_rate))

float

### Apply our solution to every cell in a column

In [107]:
df['int_rate_float'] = df['int_rate'].apply(int_rate_to_float)
df.head()

loan_amnt  funded_amnt  ...  settlement_term int_rate_float
0      35000        35000  ...              NaN          14.47
1       5000         5000  ...              NaN          22.35
2      10000        10000  ...              NaN          23.40
3      17100        17100  ...              NaN          18.94
4       4000         4000  ...              NaN          10.72

[5 rows x 141 columns]

In [0]:
# What type of data is held in our new column?

# Look at the datatypes of the last 5 columns

## Challenge

We can create a new column with our cleaned values or overwrite the original, whatever we think best suits our needs. On your assignment you will take the same approach in trying to methodically build up the complexity of your code until you have a few lines that will work for any cell in a column. At that point you'll contain all of that functionality in a reusable function block and then use the `.apply()` function to... well... apply those changes to an entire column.

# [Objective](#pandas-apply) Modify and Create Columns using `.apply()`



## Overview

We're already seen one example of using the `.apply()` function to clean up a column. Lets see if we can do it again, but this time on a slightly more complicated use case.

Remember, the goal here is to write a function that will work correctly on any **individual** cell of a specific column. Then we can reuse that function on those individual cells of a dataframe column via the `.apply()` function.

Lets clean up the emp_title "Employment Title" column!

## Follow Along

First we'll try and diagnose how bad the problem is and what improvements we might be able to make.

In [108]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False).head(20)

NaN                   20947
Teacher                2090
Manager                1773
Registered Nurse        952
Driver                  924
RN                      726
Supervisor              697
Sales                   580
Project Manager         526
General Manager         523
Office Manager          521
Owner                   420
Director                402
Truck Driver            387
Operations Manager      387
Nurse                   326
Engineer                325
Sales Manager           304
manager                 301
Supervisor              270
Name: emp_title, dtype: int64

In [109]:
# How many different unique employment titles are there currently?
len(df['emp_title'].value_counts())

43892

In [111]:
df['emp_title'].nunique()

43892

In [113]:
# How often is the employment_title null?
df['emp_title'].isnull().sum()

20947

What are some possible reasons as to why a person's employment title may have not been provided?

In [0]:
import numpy as np
# Create some examples that represent the cases that we want to clean up

examples = ['owner', 'Supervisor', 'Project Manager', np.NaN]

In [133]:
# Write a function to clean up these use cases and increase uniformity.
def clean_title(title):
  if isinstance(title, str):
    return title.strip().title()
  else:
    return "Unknowns"

for example in examples:
  print(clean_title(example))

Owner
Supervisor
Project Manager
Unknowns


In [134]:
# list comprehensions can combine function calls and for loops over lists
# into one succinct and fairly readable single line of code.

[clean_title(example) for example in examples]

['Owner', 'Supervisor', 'Project Manager', 'Unknowns']

In [0]:
# We have a function that works as expected. Lets apply it to our column.
# This time we'll overwrite the original column

df['emp_title'] = df['emp_title'].apply(clean_title)

We can use the same code as we did earlier to see how much progress was made.


In [137]:
# Look at the top 20 employment titles
df['emp_title'].value_counts(dropna=False).head(20)

Unknowns              20947
Teacher                2557
Manager                2395
Registered Nurse       1418
Driver                 1258
Supervisor             1160
Truck Driver            920
Rn                      834
Office Manager          805
Sales                   803
General Manager         791
Project Manager         720
Owner                   625
Director                523
Operations Manager      518
Sales Manager           500
Police Officer          440
Nurse                   425
Technician              420
Engineer                412
Name: emp_title, dtype: int64

In [138]:
# How many different unique employment titles are there currently?
len(df['emp_title'].value_counts())

34902

In [0]:
# How often is the employment_title null (NaN)?
df['emp_title']

## Challenge

Using the .apply() function isn't always about creating new columns on a dataframe, we can use it to clean up or modify existing columns as well. 

# [Objective](#dates-and-times) Work with Dates and Times with Pandas

## Overview

Pandas has its own datatype datatype that makes it extremely convenient to convert strings that are in standard date formates to datetime objects and then use those datetime objects to either create new features on a dataframe or work with the dataset in a timeseries fashion. 

This section will demonstrate how to take a column of date strings, convert it to a datetime object and then use the datetime formatting `.dt` to access specific parts of the date (year, month, day) to generate useful columns on a dataframe.

## Follow Along

### Work with Dates 

pandas documentation
- [to_datetime](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html)
- [Time/Date Components](https://pandas.pydata.org/pandas-docs/stable/timeseries.html#time-date-components) "You can access these properties via the `.dt` accessor"

Many of the most useful date columns in this dataset have the suffix `_d` to indicate that they correspond to dates.

We'll use a list comprehension to print them out

In [0]:
[col for col in df if col.endswith('_d')]

In [0]:
df['issue_d'].value_counts()

Lets look at the string format of the `issue_d` column

In [0]:
df['issue_d'].head()

Because this string format %m-%y is a common datetime format, we can just let Pandas detect this format and translate it to the appropriate datetime object.

In [0]:
df['issue_d'] = pd.to_datetime(df['issue_d'], infer_datetime_format=True)
df.dtypes[:15]

Now we can see that the `issue_d` column has been changed to hold `datetime` objects.

Lets look at one of the cells specifically to see what a datetime object looks like:

In [0]:
df['issue_d'][0]

You can see how the month and year have been indicated by the strings that were contained in the column previously, and that the rest of the values have been inferred.

In [0]:
df['issue_d'].head().values

We can use the `.dt` accessor to now grab specific parts of the datetime object. Lets grab just the year from the all of the cells in the `issue_d` column

In [0]:
df['issue_d'].dt.year

Now the month.

In [0]:
df['issue_d'].dt.month

It's just that easy! Now, instead of printing them out, lets add these year and month values as new columns on our dataframe. Again, you'll have to scroll all the way over to the right in the table to see the new columns.

In [0]:
df['issue_year'] = df['issue_d'].dt.year
df['issue_month'] = df['issue_d'].dt.month

df.head()

Because all of these dates come from Q4 of 2018, the `issue_d` column isn't all that interesting. Lets look at the `earliest_cr_line` column, which is also a string, but that could be converted to datetime format.

We're going to create a new column called `days_from_earliest_credit_to_issue`

It's a long column header, but think about how valuable this piece of information could be. This number will essentially indicate the length of a person's credit history and if that is correlated with repayment or other factors could be a valuable predictor!

In [0]:
df['earliest_cr_line'].head()

In [0]:
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], infer_datetime_format=True)

What we're about to do is so cool! Pandas' datetime format is so smart that we can simply use the subtraction operator `-` in order to calculate the amount of time between two dates. 

Think about everything that's going on under the hood in order to give us such straightforward syntax! Handling months of different lengths, leap years, etc. Pandas datetime objects are seriously powerful!

In [159]:
(df['issue_d'] - df['earliest_cr_line']).dt.days

0          8645
1          3197
2          4262
3          7580
4          4474
          ...  
128407    12114
128408    13241
128409     4505
128410     3652
128411     4413
Length: 128412, dtype: int64

In [0]:
df['days_from_credit_to_issue'] = (df['issue_d'] - df['earliest_cr_line']).dt.days

What's oldest credit history that was involved in Q4 2018? 

In [161]:
df['days_from_credit_to_issue'].describe()

count    128412.000000
mean       5859.891490
std        2886.535578
min        1126.000000
25%        4049.000000
50%        5266.000000
75%        7244.000000
max       25171.000000
Name: days_from_credit_to_issue, dtype: float64

25,171 days is ~ 68.96 years of credit history!

## Challenge

Pandas' datetime format is so easy to work with that there's really no excuse for not using dates to make features on a dataframe! Get ready to practice more of this on your assignment.